### PR TITLE
[ioctl] Refactor bcblock command in new ioctl

### LINKED ISSUE #3267: [ioctl] refactor bcblock command  in new ioctl
### What would you like to be added
 - Step 1: refactor  command  [newcmd/bc/bcblock.go](https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/bc/bcblock.go) 
 - Step 2: append necessary unit test case in [newcmd/bc/bcblock_test.go](https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/bc/bcblock_test.go) 

#### Notes:
1. Use [function of client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) . 
2. [Output package](https://github.com/iotexproject/iotex-core/blob/master/ioctl/output) is deprecated. `e.g.` `output.PrintError` could be replaced with `errors.Wrap`.
3. Global variables (config.ReadConfig  config.DefaultConfigFile etc.) have been wrapped into client interface. Please use [client interface](https://github.com/iotexproject/iotex-core/blob/master/ioctl/client.go) to access them if needed
4. The logic of new code should be similar to [the old one](https://github.com/iotexproject/iotex-core/blob/master/ioctl/cmd/bc/bcblock.go ).
5. In each unit test, the results should be checked whether they are as the expected. 

#### Example:
1. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountinfo.go
2. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountdelete.go
3. https://github.com/iotexproject/iotex-core/blob/master/ioctl/newcmd/account/accountcreateadd.go

### Why is this needed
Unit tests couldn't be added for the old ioctl command (./ioctl/cmd)

### PR DESCRIPTION
**Pull request by LuckyPigeon**
# Description

Refactor `bcblock` command in new `ioctl`, with the following note.

* Output package is deprecated, replace it with errors package.
* Replace fmt.Println with cmd.Println
* Refactor unit test to cover the modification.

Fixes #3267 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] TestNewBCBlockCmd

**Test Configuration**:

- Firmware version: Ubuntu 21.04
- Hardware:
- Toolchain:
- SDK:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: ioctl/newcmd/bc/bcblock.go [hunk 1]
```diff
@@ -12,6 +12,7 @@ import (
 	"strconv"
 
 	"github.com/iotexproject/iotex-core/ioctl"
+	"github.com/pkg/errors"
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcblock.go:16]: delete space line

File: ioctl/newcmd/bc/bcblock.go [hunk 2]
```diff
@@ -74,7 +75,7 @@ func NewBCBlockCmd(client ioctl.Client) *cobra.Command {
 				if err != nil {
 					isHeight = false
 				} else if err = validator.ValidatePositiveNumber(int64(height)); err != nil {
-					return output.NewError(output.ValidationError, "invalid height", err)
+					return errors.Wrap(err, "invalid height")
 				}
 			} else {
 				chainMetaResponse, err := apiServiceClient.GetChainMeta(context.Background(), &iotexapi.GetChainMetaRequest{})
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcblock.go:81]: use GetChainMeta(client)

File: ioctl/newcmd/bc/bcblock.go [hunk 3]
```diff
@@ -198,12 +185,12 @@ func getBlockMetaByHeight(cli *iotexapi.APIServiceClient, height uint64) (*iotex
 	if err != nil {
 		sta, ok := status.FromError(err)
 		if ok {
-			return nil, output.NewError(output.APIError, sta.Message(), nil)
+			return nil, errors.Wrap(nil, sta.Message())
```
Comment (CoderZhi) [inline ioctl/newcmd/bc/bcblock.go:188]: `return nil, errors.New(stat.Message())`

File: ioctl/newcmd/bc/bcblock.go [hunk 4]
```diff
@@ -166,12 +153,12 @@ func getActionInfoWithinBlock(cli *iotexapi.APIServiceClient, height uint64, cou
...
```
Comment (CoderZhi) [inline ioctl/newcmd/bc/bcblock.go:156]: `return nil, errors.New(sta.Message())`

File: ioctl/newcmd/bc/bcblock.go [hunk 5]
```diff
@@ -109,22 +110,16 @@ func NewBCBlockCmd(client ioctl.Client) *cobra.Command {
 							SenderPubKey: item.SenderPubKey,
 							Signature:    item.Signature,
 						}
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcblock.go:112]: add
https://github.com/iotexproject/iotex-core/blob/651d77bb39b759614627003f9a2441fe0ff3bb8b/ioctl/cmd/bc/bcblock.go#L177-L182

File: ioctl/newcmd/bc/bcblock.go [hunk 6]
```diff
@@ -144,14 +139,6 @@ type actionInfo struct {
...
-func (m *blockMessage) String() string {
-	if output.Format == "" {
-		message := fmt.Sprintf("Blockchain Node: %s\n%s\n%s", m.Node, output.JSONString(m.Block), output.JSONString(m.Actions))
-		return message
-	}
-	return output.FormatString(output.Result, m)
-}
-
 // getActionInfoByBlock gets action info by block hash with start index and action count
 func getActionInfoWithinBlock(cli *iotexapi.APIServiceClient, height uint64, count uint64) ([]*iotexapi.BlockInfo, error) {
```
Comment (huof6829) [inline ioctl/newcmd/bc/bcblock.go:143]: can combine getBlockMetaByHeight  and getActionInfoWithinBlock ?  because two function are similiar

**Review by huof6829 — APPROVED** [review]

File: ioctl/newcmd/bc/bcblock.go [hunk 7]
```diff
@@ -166,55 +177,19 @@ func getActionInfoWithinBlock(cli *iotexapi.APIServiceClient, height uint64, cou
 	if err != nil {
 		sta, ok := status.FromError(err)
 		if ok {
-			return nil, output.NewError(output.APIError, sta.Message(), nil)
+			return nil, errors.Wrap(nil, sta.Message())
```
Comment (CoderZhi) [inline ioctl/newcmd/bc/bcblock.go:180]: errors.Wrap(nil, ...) == nil

File: ioctl/newcmd/bc/bcblock.go [hunk 8]
```diff
@@ -166,55 +177,19 @@ func getActionInfoWithinBlock(cli *iotexapi.APIServiceClient, height uint64, cou
...
 		}
-		return nil, output.NewError(output.NetworkError, "failed to invoke GetRawBlocks api", err)
+		return nil, errors.Wrap(err, "failed to invoke GetRawBlocks api")
 	}
 	if len(response.Blocks) == 0 {
-		return nil, output.NewError(output.APIError, "no actions returned", err)
+		return nil, errors.Wrap(err, "no actions returned")
```
Comment (CoderZhi) [inline ioctl/newcmd/bc/bcblock.go:185]: err == nil

File: ioctl/newcmd/bc/bcblock.go [hunk 9]
```diff
@@ -226,12 +201,12 @@ func getBlockMetaByHash(cli *iotexapi.APIServiceClient, hash string) (*iotextype
...
 		if ok {
-			return nil, output.NewError(output.APIError, sta.Message(), nil)
+			return nil, errors.New(sta.Message())
 		}
-		return nil, output.NewError(output.NetworkError, "failed to invoke GetBlockMetas api", err)
+		return nil, errors.Wrap(err, "failed to invoke GetBlockMetas api")
 	}
 	if len(response.BlkMetas) == 0 {
-		return nil, output.NewError(output.APIError, "no block returned", err)
+		return nil, errors.Wrap(err, "no block returned")
```
Comment (CoderZhi) [inline ioctl/newcmd/bc/bcblock.go:209]: err == nil

File: ioctl/newcmd/bc/bcblock.go [hunk 10]
```diff
@@ -74,57 +75,70 @@ func NewBCBlockCmd(client ioctl.Client) *cobra.Command {
...
 					return err
 				}
-				height = chainMetaResponse.ChainMeta.Height
+				height = chainMeta.Height
 			}
 			var blockMeta *iotextypes.BlockMeta
 			var blocksInfo []*iotexapi.BlockInfo
 			if isHeight {
-				blockMeta, err = getBlockMetaByHeight(&apiServiceClient, height)
+				request := &iotexapi.GetBlockMetasRequest{
```
Comment (Liuhaai) [inline ioctl/newcmd/bc/bcblock.go:90]: create `request` in `if len(args) != 0 {} else {}` condition

File: ioctl/newcmd/bc/bcblock.go [hunk 11]
```diff
@@ -69,62 +69,81 @@ func NewBCBlockCmd(client ioctl.Client) *cobra.Command {
 				return err
 			}
 
+			var request *iotexapi.GetBlockMetasRequest
 			if len(args) != 0 {
```
Comment (Liuhaai) [inline ioctl/newcmd/bc/bcblock.go:73]: could you create a subfunc for this condition? `func parseArg(arg) request`?

File: ioctl/newcmd/bc/bcblock.go [hunk 12]
```diff
@@ -198,40 +182,58 @@ func getBlockMetaByHeight(cli *iotexapi.APIServiceClient, height uint64) (*iotex
...
+func parseArg(c ioctl.Client, arg string) (*iotexapi.GetBlockMetasRequest, error) {
+	var (
+		height  uint64
+		err     error
+		request *iotexapi.GetBlockMetasRequest
+	)
+	if arg != "" {
+		height, err = strconv.ParseUint(arg, 10, 64)
+		if err != nil {
+			request = &iotexapi.GetBlockMetasRequest{
```
Comment (Liuhaai) [inline ioctl/newcmd/bc/bcblock.go:205]: direct return

File: ioctl/newcmd/bc/bcblock.go [hunk 13]
```diff
@@ -198,40 +182,58 @@ func getBlockMetaByHeight(cli *iotexapi.APIServiceClient, height uint64) (*iotex
...
+				Lookup: &iotexapi.GetBlockMetasRequest_ByHash{
+					ByHash: &iotexapi.GetBlockMetaByHashRequest{BlkHash: arg},
+				},
+			}
+		} else if err = validator.ValidatePositiveNumber(int64(height)); err != nil {
+			return nil, errors.Wrap(err, "invalid height")
+		} else {
+			request = &iotexapi.GetBlockMetasRequest{
```
Comment (Liuhaai) [inline ioctl/newcmd/bc/bcblock.go:213]: direct return

File: ioctl/newcmd/bc/bcblock.go [hunk 14]
```diff
@@ -198,40 +182,58 @@ func getBlockMetaByHeight(cli *iotexapi.APIServiceClient, height uint64) (*iotex
...
+				Lookup: &iotexapi.GetBlockMetasRequest_ByIndex{
+					ByIndex: &iotexapi.GetBlockMetasByIndexRequest{
+						Start: height,
+						Count: 1,
+					},
+				},
+			}
+		}
+	} else {
```
Comment (Liuhaai) [inline ioctl/newcmd/bc/bcblock.go:222]: remove else in this subfunc

Comment (Liuhaai) [reply ioctl/newcmd/bc/bcblock.go]: the issues of other `else` are similar

Comment (LuckyPigeon) [reply ioctl/newcmd/bc/bcblock.go]: But their height are different

File: ioctl/newcmd/bc/bcblock.go [hunk 15]
```diff
@@ -198,40 +182,57 @@ func getBlockMetaByHeight(cli *iotexapi.APIServiceClient, height uint64) (*iotex
...
-	if err == nil {
-		ctx = metautils.NiceMD(jwtMD).ToOutgoing(ctx)
+// parseArg parse argument and returns GetBlockMetasRequest
+func parseArg(c ioctl.Client, arg string) (*iotexapi.GetBlockMetasRequest, error) {
+	var (
+		height  uint64
+		err     error
+		request *iotexapi.GetBlockMetasRequest
+	)
+	if arg != "" {
```
[the remark this answers is not in the document]
Comment (Liuhaai) [reply ioctl/newcmd/bc/bcblock.go:202]: let me know if there is any issue with this pattern

[the remark this answers is not in the document]
Comment (LuckyPigeon) [reply ioctl/newcmd/bc/bcblock.go:202]: If we move else statement out of `arg != ""`, then we will encounter invalid arg (`arg == ""`) error.

[the remark this answers is not in the document]
Comment (LuckyPigeon) [reply ioctl/newcmd/bc/bcblock.go:202]: So it's not wise to move else statement out.

[the remark this answers is not in the document]
Comment (Liuhaai) [reply ioctl/newcmd/bc/bcblock.go:202]: sorry, the comment is on L204

File: ioctl/newcmd/bc/bcblock.go [hunk 16]
```diff
@@ -198,40 +182,57 @@ func getBlockMetaByHeight(cli *iotexapi.APIServiceClient, height uint64) (*iotex
...
-		if ok {
-			return nil, output.NewError(output.APIError, sta.Message(), nil)
-		}
-		return nil, output.NewError(output.NetworkError, "failed to invoke GetBlockMetas api", err)
+		return nil, err
 	}
-	if len(response.BlkMetas) == 0 {
-		return nil, output.NewError(output.APIError, "no block returned", err)
+	height = chainMeta.Height
+	request = &iotexapi.GetBlockMetasRequest{
```
Comment (Liuhaai) [inline ioctl/newcmd/bc/bcblock.go:228]: final comment: directly return &iotexapi.GetBlockMetasRequest{}, nil

**Review by Liuhaai — APPROVED** [review]

**Pull request merged**